grpc/dispatcher.go

```diff
@@ -0,0 +1,86 @@
+package grpc
+
+import (
+	"context"
+	"time"
+
+	"github.com/gojek/fiber"
+	fiberError "github.com/gojek/fiber/errors"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/credentials/insecure"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/proto"
+)
+
+const (
+	TimeoutDefault = time.Second
+)
+
+type Dispatcher struct {
+	Timeout time.Duration
+}
+
+func (d *Dispatcher) Do(request fiber.Request) fiber.Response {
+
+	grpcRequest, ok := request.(*Request)
+	if !ok {
+		return fiber.NewErrorResponse(
+			fiberError.FiberError{
+				Code:    int(codes.InvalidArgument),
+				Message: "fiber: grpc.Dispatcher supports only grpc.Request type of requests",
+			})
+	}
+
+	if grpcRequest.hostport == "" || grpcRequest.ServiceMethod == "" {
+		return fiber.NewErrorResponse(
+			fiberError.FiberError{
+				Code:    int(codes.InvalidArgument),
+				Message: "missing hostport/servicemethod",
+			})
+	}
+
+	conn, err := grpc.Dial(grpcRequest.hostport, grpc.WithTransportCredentials(insecure.NewCredentials()))
+	if err != nil {
+		// if ok is false, unknown codes.Unknown and Status msg is returned in Status
+		responseStatus, _ := status.FromError(err)
+		return &Response{Status: *responseStatus}
+	}
+
+	payload, ok := request.Payload().(proto.Message)
+	if !ok {
+		return fiber.NewErrorResponse(
+			fiberError.FiberError{
+				Code:    int(codes.InvalidArgument),
+				Message: "unable to convert payload to proto message",
+			})
+	}
+
+	// Set default timeout
+	if d.Timeout == 0 {
+		d.Timeout = TimeoutDefault
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), d.Timeout)
+	defer cancel()
```

Comment:
I wonder if it's necessary, since at this point the call has been completed and hence doesn't matter if we cancel the context or not.

Reply:
just found out for grpc it should be a `dialOption` and passed while `.dial` 

so these arent necessary, will refactor them.

Reply:
I realize i misunderstood the comment initially, there is no need to cancel.

The dialOption [withTimeout](https://github.com/grpc/grpc-go/blob/v1.49.0/dialoptions.go#L353) is deprecating, will just follow [this](https://stackoverflow.com/a/63969292). Its weird cause https://github.com/grpc/grpc-go/blob/master/clientconn.go#L217 is still expecting `withTimeout` implementation of having the dialoption apply. 

I looked through the possibility of mapping dial options from config, however we would need to supply the corresponding mapping function for the `apply` function of `dialOption`. Will add this as requirements come.

Reply:
actually cant remove this, as it will fail lint lol